AOTInductor dynamic shape

chenyang78 opened the pull request:
Summary: This PR adds dynamic-shape support for AOTInductor

* On the runtime/interface side, we added two structs, StaticDimInfo
and DynamicDimInfo, to hold values for static and dynamic dimensions,
respectively. Dynamic dimensions are tracked by an unordered map field
defined in AOTInductorModelBase. At inference time, the inference run
method will assign the current real dimensional value to each dynamic
dimension before executing any kernel.

* On the CUDA wrapper codegen side, we generate dynamic symbols
appropriately for shape computations. We simulate kernel launch grids
in the C++ land by re-using the grid functions from the Python world.
The returned grid configs, which may contain symbolic expressions,
are printed out in their C++ forms via the CppPrinter. Note that
when dynamic shapes are involved, we have to compute grid configs
for each kernel at runtime in the same way as we do for launching
the corresponding Triton kernel. Otherwise, we may end up with
memory-access failures or mis-computations caused by invalid indices
for fetching or storing data in device memory.

Differential Revision: D49100472




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @kadeng @muchulee8 @aakhundov

ipiszy commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1708,6 +1759,23 @@ def generate_args_decl(self, call_args):
 
         return ", ".join(new_args)
 
+    def generate_default_grid(self, name, grid, cuda=True):
```
Is this cuda arg used anywhere?

desertfire replied:
Seems unnecessary since the cpu backend doesn't need to generate grid

ipiszy commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1708,6 +1759,23 @@ def generate_args_decl(self, call_args):
+            return grid
+        assert isinstance(grid, list), f"expected {grid=} to be a list"
+        grid = [e.inner_expr if isinstance(e, SymbolicCallArg) else e for e in grid]
+        grid_fn = default_grid(*grid)
```
What is `default_grid`? I cannot find it in the pytorch repo.

desertfire replied:
It's from `from ..triton_heuristics import grid as default_grid`

ipiszy commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -732,6 +732,9 @@ def generate_profiler_mark_wrapper_call(self, stack):
         )
         stack.enter_context(self.wrapper_call.indent())
 
+    def generate_default_grid(self, name, grid_args):
```
nit: Add more comments to describe what this function does, and the args. Also add type hints.

chenyang78 replied:
Will do. Thanks.

ipiszy commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -115,6 +116,7 @@ def get_cpp_op_schema(kernel):
 @dataclasses.dataclass
 class SymbolicCallArg:
     inner: Any
+    inner_expr: sympy.Expr
```
Add some comments?

chenyang78 replied:
Done.

ipiszy reviewed (commented):
Thanks @chenyang78 ! Left some comments, mainly some clarification questions. It would be great if you could help summarize your changes in the PR description.

desertfire commented on torch/csrc/inductor/aot_runtime/interface.h:
```diff
@@ -66,6 +66,7 @@ AOTInductorError AOTInductorModelContainerRun(
     size_t num_inputs,
     AOTInductorTensorHandle outputs_handle,
     size_t num_outputs,
+    AOTInductorParamShape* output_shapes,
```
Needs to revisit for ABI compatibility, but ok for now.

desertfire reviewed (approved):
Left some nits. LGTM overall.

Currently AOTInductor on the OSS benchmarks are not a part of the CI yet. I have a pending PR, https://github.com/pytorch/pytorch/pull/108419, but it won't be merged soon given the recent CI infra instability. So can you manually trigger a run at https://github.com/pytorch/pytorch/actions/workflows/inductor-perf-test-nightly.yml?

You will need to select the options like in the screenshot, 
![Screenshot 2023-09-12 at 10 33 33 PM](https://github.com/pytorch/pytorch/assets/3659962/1818946e-1557-43bd-8afa-938f1a9ce852)
Unfortunately the Branch selection does not work with a PR from your forked repo. You can create a mirror PR on a pytorch branch in order to do the testing. Let me know if you need more help on the instructions.

chenyang78 commented on test/inductor/test_aot_inductor.py:
```diff
@@ -47,17 +52,57 @@ def load(cls, model, example_inputs, example_outputs, options=None):
+        for example_inputs, example_outputs in zip(
+            list_example_inputs, list_example_outputs
+        ):
+            output_tensors = [torch.empty_like(output) for output in example_outputs]
```
No, we don't re-use output tensors. Instead, we allocate output tensors for each inference run and return these tensors to the caller of the forward function.

chenyang78 commented:
> Left some nits. LGTM overall.
> 
> Currently AOTInductor on the OSS benchmarks are not a part of the CI yet. I have a pending PR, #108419, but it won't be merged soon given the recent CI infra instability. So can you manually trigger a run at https://github.com/pytorch/pytorch/actions/workflows/inductor-perf-test-nightly.yml?
> 
> You will need to select the options like in the screenshot, ![Screenshot 2023-09-12 at 10 33 33 PM](https://user-images.githubusercontent.com/3659962/267516871-1818946e-1557-43bd-8afa-938f1a9ce852.png) Unfortunately the Branch selection does not work with a PR from your forked repo. You can create a mirror PR on a pytorch branch in order to do the testing. Let me know if you need more help on the instructions.

Running it: https://github.com/pytorch/pytorch/actions/runs/6170063555

Hopefully I was doing it right. Thanks.

desertfire commented:
> > Left some nits. LGTM overall.
> > Currently AOTInductor on the OSS benchmarks are not a part of the CI yet. I have a pending PR, #108419, but it won't be merged soon given the recent CI infra instability. So can you manually trigger a run at https://github.com/pytorch/pytorch/actions/workflows/inductor-perf-test-nightly.yml?
> > You will need to select the options like in the screenshot, ![Screenshot 2023-09-12 at 10 33 33 PM](https://user-images.githubusercontent.com/3659962/267516871-1818946e-1557-43bd-8afa-938f1a9ce852.png) Unfortunately the Branch selection does not work with a PR from your forked repo. You can create a mirror PR on a pytorch branch in order to do the testing. Let me know if you need more help on the instructions.
> 
> Running it: https://github.com/pytorch/pytorch/actions/runs/6170063555
> 
> Hopefully I was doing it right. Thanks.

Given the long task queueing time, I ran this PR locally to verify. I haven't checked every model, but the results look good to me. I am ok with landing it.

Commit:
AOTInductor dynamic shape (#109012)

Summary:

This PR adds dynamic-shape support for AOTInductor

Reviewed By: frank-wei, khabinov

Differential Revision: D49100472

chenyang78 commented:
@pytorchbot merge
